src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -50,7 +50,11 @@ import           System.FS.BlockIO.API (HasBlockIO)
 -- Since we always resolve all entries of the same key in one go, there is no
 -- need to store incompletely-resolved entries.
 data Merge t m h = Merge {
-      mergeType       :: !t
+      narrowMergeType :: !t
+      -- | We also store @toMergeType narrowMergeType@ here to avoid recomputing
+      -- it again and again for each call to 'steps', which would also add
+      -- an 'IsMergeType' constraint to large parts of the interface.
+    , mergeType       :: !MergeType
```

Comment:
Ah! Yes, exactly what I was suggesting!

I agree. Can we squash this patch into the previous one? That'd remove the add/remove noise.

Reply:
I squashed this patch into the previous one, and also did a little more squashing. I decided to keep the "refactor snapshot encoding of merge types" commit separate, though.